browser/caliban.py

```diff
@@ -473,6 +475,110 @@ def action_threshold(self, y1, x1, y2, x2, label):
         if np.any(np.isin(self.frame[..., self.feature], label)):
             self.add_cell_info(add_label=label, frame=self.frame_id)
 
+    def action_active_contour(self, label):
+        label_img = np.copy(self.frame[..., self.feature])
+
+        # get centroid of selected label
+        props = regionprops(np.where(label_img == label, label, 0))[0]
+
+        # make bounding box size to encompass some background
+        box_height = props['bbox'][2] - props['bbox'][0]
+        y1 = max(0, props['bbox'][0] - box_height // 2)
+        y2 = min(self.project.height, props['bbox'][2] + box_height // 2)
+
+        box_width = props['bbox'][3] - props['bbox'][1]
+        x1 = max(0, props['bbox'][1] - box_width // 2)
+        x2 = min(self.project.width, props['bbox'][3] + box_width // 2)
+
+        # relevant region of label image to work on
+        label_img = label_img[y1:y2, x1:x2]
+
+        # use existing label as initial level set for contour calculations
+        level_set = np.where(label_img == label, 1, 0)
+
+        # normalize input 2D frame data values to range [0.0, 1.0]
+        adjusted_raw_frame = Normalize()(self.raw_frame[..., self.channel])
+        predict_area = adjusted_raw_frame[y1:y2, x1:x2]
+        # import pdb; pdb.set_trace()
+
+        # returns 1 where label is predicted to be based on contouring, 0 background
+        contoured = morphological_chan_vese(predict_area, 100, init_level_set=level_set)
+
+        # contoured area should get original label value
+        contoured_label = np.where(contoured, label, 0)
+        # contours tend to fit very tightly, a small expansion here works well
+        contoured_label = dilation(contoured_label, disk(3))
+
+        # don't want to leave the original (un-contoured) label in the image
+        label_img = np.where(label_img == label, 0, label_img)
+
+        # never overwrite other labels with new contoured label
+        safe_overlay = np.where(label_img == 0, contoured_label, label_img)
```

Comment:
Are these the same commands? I think the single `np.where` is faster.

```suggestion
        # don't want to leave the original (un-contoured) label in the image
        # never overwrite other labels with new contoured label
        cond = np.logical_or(label_img == label, label_img == 0)
        safe_overlay = np.where(cond, contoured_label, label_img)
```

Reply:
I didn't find a difference between these with timeit, but I think this logic is easier to follow